### PR TITLE
Move check of invalid use of aggregate vars

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

I think the check for reuse of variables assignent in COLLECT AGGREGATE has to be moved into a seperate check function

- [x] :hankey: Bugfix 
#### Backports:

- [x] Backports required for: 3.7

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added **Regression Tests**
  - [ ] Added new C++ **Unit Tests**
  - [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/12272/

### CODE REVIEW

**Comment by jsteemann** [comment]
Seems to work, but produces a test failure here due to an error code mismatch:
```
* Test "shell_server_aql_4"
    [FAILED]  tests/js/server/aql/aql-optimizer-collect-aggregate.js

      "testAggregateAllReferToCollectvariable" failed: Error: at assertion #9: unexpected error code (AQL: use of COLLECT variable 'length1' inside same COLLECT near 'RETURN doc' at position 1:103 (while parsing) while executing: 'FOR j IN UnitTestsCollection COLLECT doc = j AGGREGATE length1 = LENGTH(1), length2 = LENGTH(length1) RETURN doc' expecting: 1512): : (1574) is not equal to (1512)
(Error
    at assertEqual (/work/ArangoDB/js/common/modules/jsunity/jsunity.js:108:19)
    at assertQueryError (/work/ArangoDB/js/server/modules/@arangodb/aql-helper.js:154:5)
    at Object.testAggregateAllReferToCollectvariable (tests/js/server/aql/aql-optimizer-collect-aggregate.js:432:7)
    at Object.run (/work/ArangoDB/js/common/modules/jsunity/jsunity.js:549:27)
    at Object.Run [as run] (/work/ArangoDB/js/common/modules/jsunity.js:276:24)
    at tests/js/server/aql/aql-optimizer-collect-aggregate.js:1976:9
    at tests/js/server/aql/aql-optimizer-collect-aggregate.js:1985:3
    at RunTest (/work/ArangoDB/js/common/modules/jsunity.js:390:12)
    at source (eval at <anonymous> (unknown source), <anonymous>:2:52))
    at Object.run (/work/ArangoDB/js/common/modules/jsunity/jsunity.js:569:23)
    at Object.Run [as run] (/work/ArangoDB/js/common/modules/jsunity.js:276:24)
    at tests/js/server/aql/aql-optimizer-collect-aggregate.js:1976:9
    at tests/js/server/aql/aql-optimizer-collect-aggregate.js:1985:3
    at RunTest (/work/ArangoDB/js/common/modules/jsunity.js:390:12)
    at source (eval at <anonymous> (unknown source), <anonymous>:2:52) - test failed

   Suites failed: 1 Tests Failed: 1
```

File: arangod/Aql/grammar.y [hunk 1]
```diff
@@ -129,32 +115,47 @@ void registerAssignVariables(Parser* parser, arangodb::aql::Scopes* scopes,
...
-        return false;
+      
+      // check if any of the assignment refers to a variable introduced by this very
+      // same COLLECT, e.g. COLLECT aggregate x = .., y = x
+      varsInAssignment.clear();
+      Ast::getReferencedVariables(member->getMember(1), varsInAssignment);
+      for (auto const& it : varsInAssignment) {
+        if (variablesIntroduced.find(it) != variablesIntroduced.end()) {
+          std::string msg("use of COLLECT variable '" + it->name + "' inside same COLLECT");
+          parser->registerParseError(TRI_ERROR_QUERY_INVALID_AGGREGATE_EXPRESSION, msg.c_str(), it->name.c_str(), line, column);
```
Comment (jsteemann) [inline arangod/Aql/grammar.y:156]: Ì think previously we returned a different error here (`TRI_ERROR_QUERY_VARIABLE_NAME_UNKNOWN`).
That's probably also why the one test fails.

**Review by jsteemann — APPROVED** [review]
LGTM except the error code change/test failure.
I am happy if we change the error code returned by the validation, but in this case we obviously need to adjust the test case.

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/12288/

**Comment by graetzer** [comment]
I changed the error code back, I would have preferred the other one but lets stay consistent

**Pull request merged**